File: inventory_store.py
import json
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional
from uuid import uuid4

from ci_utils import build_export_payload, parse_ci_notes


INVENTORY_DIR = Path.cwd() / "data"
INVENTORY_PATH = INVENTORY_DIR / "inventory.json"


def now_iso() -> str:
    return datetime.now().isoformat(timespec="seconds")


def ensure_inventory_file() -> None:
    INVENTORY_DIR.mkdir(exist_ok=True)
    if not INVENTORY_PATH.exists():
        INVENTORY_PATH.write_text(json.dumps({"records": []}, indent=2), encoding="utf-8")


def load_records() -> List[Dict[str, Any]]:
    ensure_inventory_file()
    payload = json.loads(INVENTORY_PATH.read_text(encoding="utf-8"))
    records = payload.get("records", [])
    return sorted(records, key=lambda item: item.get("updated_at", ""), reverse=True)


def save_records(records: List[Dict[str, Any]]) -> None:
    ensure_inventory_file()
    INVENTORY_PATH.write_text(json.dumps({"records": records}, indent=2), encoding="utf-8")


def get_record(record_id: str) -> Optional[Dict[str, Any]]:
    for record in load_records():
        if record.get("id") == record_id:
            return record
    return None


def create_ci_submission(notes: str, user: Dict[str, str]) -> Dict[str, Any]:
    parsed = parse_ci_notes(notes)
    payload = build_export_payload(parsed)
    created_at = now_iso()
    record = {
        "id": uuid4().hex[:10].upper(),
        "status": "submitted_to_ao",
        "template_key": "",
        "template_label": "",
        "ci_notes": notes.strip(),
        "parsed_data": parsed,
        "account_name": parsed.get("account_name", ""),
        "applicant_name": payload.get("applicant_name", ""),
        "ci_submitter_email": user.get("email", ""),
        "ci_submitter_name": user.get("name", "CI User"),
        "ci_submitted_at": created_at,
        "ao_email": "",
        "ao_name": "",
        "ao_generated_at": "",
        "submitted_to_officer_at": "",
        "generated_file_name": "",
        "generated_file_path": "",
        "officer_note": "",
        "created_at": created_at,
        "updated_at": created_at,
    }
    records = load_records()
    records.append(record)
    save_records(records)
    return record


def update_record(record_id: str, updates: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    records = load_records()
    updated_record: Optional[Dict[str, Any]] = None
    for index, record in enumerate(records):
        if record.get("id") != record_id:
            continue
        merged = dict(record)
        merged.update(updates)
        merged["updated_at"] = now_iso()
        records[index] = merged
        updated_record = merged
        break

    if updated_record is None:
        return None

    save_records(records)
    return updated_record


def list_records_by_status(statuses: List[str]) -> List[Dict[str, Any]]:
    wanted = set(statuses)
    return [record for record in load_records() if record.get("status") in wanted]


def list_records_for_user(role: str, email: str) -> List[Dict[str, Any]]:
    email = (email or "").lower()
    records = load_records()
    if role == "ci":
        return [record for record in records if record.get("ci_submitter_email", "").lower() == email]
    if role == "ao":
        return [record for record in records if record.get("status") in {"submitted_to_ao", "generated_by_ao", "submitted_to_officer"}]
    if role == "officer":
        return [record for record in records if record.get("status") == "submitted_to_officer"]
    return records

File: php_app/src/InventoryStore.php
<?php

declare(strict_types=1);

namespace App;

final class InventoryStore
{
    public static function ensureStorage(): void
    {
        if (!is_dir(Config::dataPath())) {
            mkdir(Config::dataPath(), 0777, true);
        }

        if (!is_dir(Config::exportPath())) {
            mkdir(Config::exportPath(), 0777, true);
        }

        if (!is_file(Config::inventoryPath())) {
            file_put_contents(
                Config::inventoryPath(),
                json_encode(['records' => []], JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES)
            );
        }
    }

    public static function nowIso(): string
    {
        return date('c');
    }

    public static function loadRecords(): array
    {
        self::ensureStorage();
        $payload = json_decode((string) file_get_contents(Config::inventoryPath()), true);
        $records = $payload['records'] ?? [];
        usort(
            $records,
            static fn (array $a, array $b): int => strcmp((string) ($b['updated_at'] ?? ''), (string) ($a['updated_at'] ?? ''))
        );

        return $records;
    }

    public static function saveRecords(array $records): void
    {
        self::ensureStorage();
        file_put_contents(
            Config::inventoryPath(),
            json_encode(['records' => array_values($records)], JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES)
        );
    }

    public static function getRecord(string $recordId): ?array
    {
        foreach (self::loadRecords() as $record) {
            if (($record['id'] ?? '') === $recordId) {
                return $record;
            }
        }

        return null;
    }

    public static function createCiSubmission(string $notes, array $user, array $parsed): array
    {
        $createdAt = self::nowIso();
        $record = [
            'id' => strtoupper(substr(bin2hex(random_bytes(8)), 0, 10)),
            'status' => 'submitted_to_ao',
            'template_key' => '',
            'template_label' => '',
            'ci_notes' => trim($notes),
            'parsed_data' => $parsed,
            'account_name' => $parsed['account_name'] ?? '',
            'applicant_name' => NotesParser::joinName(
                $parsed['subject_last_name'] ?? '',
                $parsed['subject_first_name'] ?? '',
                $parsed['subject_middle_name'] ?? ''
            ),
            'ci_submitter_email' => strtolower((string) ($user['email'] ?? '')),
            'ci_submitter_name' => $user['name'] ?? 'CI User',
            'ci_submitted_at' => $createdAt,
            'ao_email' => '',
            'ao_name' => '',
            'ao_generated_at' => '',
            'submitted_to_officer_at' => '',
            'generated_file_name' => '',
            'generated_file_path' => '',
            'created_at' => $createdAt,
            'updated_at' => $createdAt,
        ];

        $records = self::loadRecords();
        $records[] = $record;
        self::saveRecords($records);

        return $record;
    }

    public static function updateRecord(string $recordId, array $updates): ?array
    {
        $records = self::loadRecords();
        $updated = null;

        foreach ($records as $index => $record) {
            if (($record['id'] ?? '') !== $recordId) {
                continue;
            }

            $updated = array_merge($record, $updates, ['updated_at' => self::nowIso()]);
            $records[$index] = $updated;
            break;
        }

        if ($updated === null) {
            return null;
        }

        self::saveRecords($records);

        return $updated;
    }

    public static function listRecordsForUser(string $role, string $email): array
    {
        $email = strtolower(trim($email));
        $records = self::loadRecords();

        return match ($role) {
            'ci' => array_values(array_filter(
                $records,
                static fn (array $record): bool => strtolower((string) ($record['ci_submitter_email'] ?? '')) === $email
            )),
            'ao' => array_values(array_filter(
                $records,
                static fn (array $record): bool => in_array((string) ($record['status'] ?? ''), ['submitted_to_ao', 'generated_by_ao', 'submitted_to_officer'], true)
            )),
            'officer' => array_values(array_filter(
                $records,
                static fn (array $record): bool => ($record['status'] ?? '') === 'submitted_to_officer'
            )),
            default => $records,
        };
    }
}
